File: Ros_Study/js_turtle_teleop/scripts/js_turtle_teleop_node.py
#!/usr/bin/env python

# -*- coding: utf-8 -*-
# @Author: IBNBlank
# @Date:   2019-01-28 22:50:17
# @Last Modified by:   IBNBlank
# @Last Modified time: 2019-01-29 22:14:20

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Joy


class Teleop(object):
    def __init__(self):
        self.__pub = rospy.Publisher('turtle1/cmd_vel', Twist, queue_size=1)
        self.__sub = rospy.Subscriber("joy", Joy, self.callback, queue_size=10)
        self.__max_lin_speed = rospy.get_param('~max_angular_speed', 2.0)
        self.__max_ang_speed = rospy.get_param('~max_linear_speed', 2.0)
        self.__lin_speed = 0
        self.__ang_speed = 0
        self.__lin_axis = rospy.get_param('~linear_axis_num', 1)
        self.__ang_axis = rospy.get_param('~angular_axis_num', 3)

    def pub_twist(self):
        twist = Twist()
        twist.linear.x = self.__lin_speed
        twist.angular.z = self.__ang_speed
        self.__pub.publish(twist)

    def callback(self, data):
        self.__lin_speed = data.axes[self.__lin_axis] * self.__max_lin_speed
        self.__ang_speed = data.axes[self.__ang_axis] * self.__max_ang_speed
        rospy.loginfo("Turtle speed change: linear:{0:.3f} angular:{1:.3f}".format(self.__lin_speed, self.__ang_speed))



if __name__ == '__main__':
    rospy.init_node('js_turtle_teleop_node', anonymous=True)
    teleop = Teleop()
    rate = rospy.Rate(50)

    while not rospy.is_shutdown():
        teleop.pub_twist()
        rate.sleep()

File: Ros_Study/js_turtle_teleop/src/js_turtle_teleop_node.cpp
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "sensor_msgs/Joy.h"

class Teleop
{
    public:
        Teleop();
        void pub_twist();

    private:
        void callback(const sensor_msgs::Joy::ConstPtr& Joy);
        ros::NodeHandle nh_;
        ros::Subscriber sub_;
        ros::Publisher pub_;
        double max_lin_speed_, max_ang_speed_, lin_speed_, ang_speed_;
        int lin_axis_, ang_axis_;
};

Teleop::Teleop(): nh_("~"), lin_speed_(0.0), ang_speed_(0.0)
{
    nh_.param<int>("linear_axis_num", lin_axis_, 1);
    nh_.param<int>("angular_axis_num", ang_axis_, 3);
    nh_.param<double>("max_linear_speed", max_lin_speed_, 2.0);
    nh_.param<double>("max_angular_speed", max_ang_speed_, 2.0);
    pub_ = nh_.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 1);
    sub_ = nh_.subscribe<sensor_msgs::Joy>("/joy", 10, &Teleop::callback, this);
}

void Teleop::pub_twist()
{
    geometry_msgs::Twist v;
    v.linear.x = lin_speed_;
    v.angular.z = ang_speed_;
    pub_.publish(v);
}

void Teleop::callback(const sensor_msgs::Joy::ConstPtr& Joy)
{
    lin_speed_ = Joy -> axes[lin_axis_] * max_lin_speed_;
    ang_speed_ = Joy -> axes[ang_axis_] * max_ang_speed_;
    ROS_INFO("Turtle speed change: linear:%.3lf angular:%.3lf", lin_speed_, ang_speed_);
}



int main(int argc,char** argv)
{
    ros::init(argc, argv, "js_turtle_teleop_node");
    Teleop teleop;
    ros::Rate loop_rate(50);

    while(ros::ok())
    {
        teleop.pub_twist();

        ros::spinOnce();
        loop_rate.sleep();
    }

    return 0;
}
